pkg/storage/chunk/storage/caching_index_client.go

```diff
@@ -55,26 +68,49 @@ type cachingIndexClient struct {
 	limits              StoreLimits
 	logger              log.Logger
 	disableBroadQueries bool
+	asyncQueue          chan cacheEntry
+	maxAsyncConcurrency int
+	maxAsyncBufferSize  int
+	stop                chan struct{}
 }
 
-func newCachingIndexClient(client chunk.IndexClient, c cache.Cache, validity time.Duration, limits StoreLimits, logger log.Logger, disableBroadQueries bool) chunk.IndexClient {
+func newCachingIndexClient(client chunk.IndexClient, c cache.Cache, validity time.Duration, limits StoreLimits, logger log.Logger, disableBroadQueries bool, maxAsyncConcurrency int, maxAsyncBufferSize int) chunk.IndexClient {
 	if c == nil || cache.IsEmptyTieredCache(c) {
 		return client
 	}
 
-	return &cachingIndexClient{
+	cacheClient := &cachingIndexClient{
 		IndexClient:         client,
 		cache:               cache.NewSnappy(c, logger),
 		validity:            validity,
 		limits:              limits,
 		logger:              logger,
 		disableBroadQueries: disableBroadQueries,
+		maxAsyncConcurrency: maxAsyncConcurrency,
+		maxAsyncBufferSize:  maxAsyncBufferSize,
+		stop:                make(chan struct{}),
+	}
+	cacheClient.asyncQueue = make(chan cacheEntry, cacheClient.maxAsyncBufferSize)
+	for i := 0; i < cacheClient.maxAsyncConcurrency; i++ {
+		go cacheClient.asyncWriteBackCacheQueueProcessLoop()
 	}
+
+	return cacheClient
+}
+
+type IndexAsyncClient interface {
+	AsyncQueueLength() int
+}
+
+type cacheEntry struct {
+	keys    []string
+	batches []ReadBatch
 }
 
 func (s *cachingIndexClient) Stop() {
 	s.cache.Stop()
 	s.IndexClient.Stop()
+	close(s.stop)
```

Comment:
After closing, should we wait for the queue to be emptied before `Stop()` returns?